NASA Challenge_[@elementrobotics]_[LunarSim]

mazm0002 opened the pull request:
# LunarSim
![LunarSim](https://github.com/user-attachments/assets/fc0858a1-36e7-4c3e-afe7-ece1d99a10e0)

Resolves #50
This PR adds a Gazebo Garden based lunar world demo to spaceROS. The features included in this PR are: 

- A world with a lunar terrain that is derived from a Digital Elevation Model (DEM) of the area around the Shackleton Crater (close to the South Pole) taken from the Lunar Orbiter Laser Altimeter (LOLA).
-  A dynamic model of the Sun that accurately depicts the low light conditions near the lunar south pole by using ephemeris data from NASA's Horizons application.
-   Improved lunar surface texture created by blending surface normals with high resolution sand texture.

This package serves as a foundation for creating a realistic lunar simulation that the community can further expand upon. To support this, we have provided comprehensive documentation that guides users through the process of preparing and adding their own DEMs, ephemeris data, and custom textures. 

This PR also depends on our PR in the spaceROS Docker repo, which adds a Docker image for the demo. This image builds on top of the spaceROS base image and follows the format used in other demos for consistency.

ivanperez-keera reviewed (changes requested):
@mazm0002
 
This PR puts a lot of data in the demos repo (156000 lines more according to Github, and several megabytes worth of data). I believe some of that would belong in the https://github.com/space-ros/simulation/ repo, since it's models and such.

Can you please take a look at that repo, see how other demos are structured, see what would make sense to store there from your submission, send a PR to that repo, and adjust this one accordingly?

Thanks!

ivanperez-keera commented:
Hi @mazm0002 Just following up on this.

mazm0002 commented:
Hi @ivanperez-keera, apologies for the delay. I've moved the world assets to the simulation repo and made a PR there https://github.com/space-ros/simulation/pull/39. I'm pulling the assets into the demo docker by adding the fork of the simulation repo to the demo_manual_pkgs.repos file. This means that once the assets have been merged, the demo_manual_pkgs.repos file needs to be updated in this PR to point to the upstream simulation repo. Let me know if there's a better way to go about this.

ivanperez-keera commented:
Thanks, @mazm0002 . Is there a way to make the screenshot image smaller? It's more than 1MB right now. Since it's meant to be read in the README, I'm hoping it'd be possible to make it smaller in size and compressing it a bit (perhaps making it a JPG?) without visible loss of precision.

mazm0002 commented:
@ivanperez-keera I've converted the image to a JPG and it's down to 426 kb now.

mazm0002 commented:
@mkhansenbot @ivanperez-keera Any idea why this is failing in CI now with this error?  `No space left on device : '/home/runner/runners/2.323.0/_diag/blocks/31686724-11e7-446f-a5d8-b04fa8a96898_c28fca0e-2bd9-5b57-8190-cdfd8e08d3b3.3'`. Only change I made since the previous commit was the conversion of an image from png to jpg.

ivanperez-keera commented:
@mazm0002 @asimonov  The CI build errors have been fixed now, but the `docker` repo doesn't build `space_robots` currently.

EzraBrooks commented:
Is this based on the LunarSim project out of PUT, or a different thing? I don't want any confusion between the (possibly many) different things called "LunarSim" in our demos

mazm0002 commented:
> Is this based on the LunarSim project out of PUT, or a different thing? I don't want any confusion between the (possibly many) different things called "LunarSim" in our demos
This demo isn't associated with the one from PUT, but I wouldn't mind changing the name to avoid any confusion with other demos. Open to any suggestions, maybe something like "lunar_rover" or "lunar_terrain"?

EzraBrooks commented:
either of those names sound good to me!

ivanperez-keera commented:
I like `lunar_rover`. Seems very descriptive.

EzraBrooks commented:
@ivanperez-keera it doesn't seem like Copilot picked up on what might be wrong in the CMake (at least in this first pass) but it did seem to generate some other helpful review comments.

ivanperez-keera commented:
I think this solves the problem:

```diff
diff --git a/lunarsim_demo/demo_manual_pkgs.repos b/lunarsim_demo/demo_manual_pkgs.repos
index 50bcd76..679088e 100644
--- a/lunarsim_demo/demo_manual_pkgs.repos
+++ b/lunarsim_demo/demo_manual_pkgs.repos
@@ -10,7 +10,7 @@ repositories:
   vision_msgs:
     type: git
     url: https://github.com/ros-perception/vision_msgs.git
-    version: ros2
+    version: humble
```

EzraBrooks commented on lunarsim_demo/lunarsim_gz_worlds/launch/display.launch.py:
```diff
@@ -0,0 +1,93 @@
+            DeclareLaunchArgument(
+                "gz_verbosity",
+                default_value="3",
+                description="Verbosity level for Ignition Gazebo (0~4).",
```
I think this could probably just be "Gazebo" - it's not called Ignition anymore, right?
```suggestion
                description="Verbosity level for Gazebo (0~4).",
```

mazm0002 commented:
@ivanperez-keera Good catch, I've updated the repo with that fix, the GPG fix, as well as the other issues Copilot brought up. I've renamed the project to lunar_terrain since I tend to agree with @asimonov that the primary addition here is the terrain. The rover model is not exactly a lunar rover, just an open source one that was accessible. I've updated the simulation and docs pull requests to reflect these changes as well.

ivanperez-keera commented:
@mkhansenbot how are these related?

I have a commit in my work machine when I built everything completely. I run into issues with keys and probably did something similar to what you did in the ROS trick demo, if that's what you are talking about.

mazm0002 commented:
Build failed since the GPG key fix for the other demos hadn't been included. I've rebased the branch, builds fine on my local now, could you rerun the build @mkhansenbot? 🤞

mazm0002 commented:
Just an fyi, https://github.com/space-ros/simulation/pull/39 should be merged in before this one, and then this PR needs to be updated to point to the main repo instead of our fork in lunar_terrain/demo_manual_pkgs.repos

ivanperez-keera commented:
For what is worth, I had to make these changes to make things work on my machine. These may not work on all machines:


```diff
$ git diff
diff --git a/lunar_terrain/run.sh b/lunar_terrain/run.sh
index ae941ae..a580b1a 100755
--- a/lunar_terrain/run.sh
+++ b/lunar_terrain/run.sh
@@ -12,5 +12,6 @@ IMG_NAME=openrobotics/space_robots_lunar_terrain
 CONTAINER_NAME="$(tr '/' '_' <<< "$IMG_NAME")"
 
 # Start the container
-docker run --rm -it --name "$CONTAINER_NAME"  --network host \
+docker run -v /dev/dri:/dev/dri --rm -it --name "$CONTAINER_NAME"  --network host \
+    -v /tmp/.X11-unix:/tmp/.X11-unix \
```

ivanperez-keera commented:
Tagging @mazm0002 for awareness.

mazm0002 commented:
> We evaluated this and it seems good to merge except for the question about the changelog indicated above:
> 
> Does the changelog, which seems to refer to a different project, need to be kept. If so, does anything in it need to be changed.
> 
> Does anybody need to be credited (e.g., if your starting point was a different package that you used as template and modified).

@ivanperez-keera IIRC we did use some code from the leo_simulator library as a template, but have kept their (Fictional) copyright license notice on all the files that were derived from theirs. The changelog doesn't need to be kept, I will remove it.

ivanperez-keera commented:
Before merging this, we need to adjust the URL to the simulator repo. It was pointing to their branch.

That commit should be squashed with the current commit.
